File: tests/fixtures/go-gin/utils/validation.go
// utils/validation.go - Input validation utilities

package utils

import (
	"regexp"
	"strings"
	"unicode"
)

var (
	emailRegex    = regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
	usernameRegex = regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9_-]*$`)
	phoneRegex    = regexp.MustCompile(`^\+?[0-9]{10,15}$`)
	slugRegex     = regexp.MustCompile(`^[a-z0-9]+(?:-[a-z0-9]+)*$`)
)

type ValidationErrors map[string][]string

func (v ValidationErrors) Add(field, message string) {
	v[field] = append(v[field], message)
}

func (v ValidationErrors) HasErrors() bool {
	return len(v) > 0
}

func ValidateEmail(email string) []string {
	var errors []string

	if email == "" {
		errors = append(errors, "email is required")
		return errors
	}

	if !emailRegex.MatchString(email) {
		errors = append(errors, "invalid email format")
	}

	return errors
}

func ValidateUsername(username string) []string {
	var errors []string

	if username == "" {
		errors = append(errors, "username is required")
		return errors
	}

	if len(username) < 3 {
		errors = append(errors, "username must be at least 3 characters")
	}

	if len(username) > 30 {
		errors = append(errors, "username must be at most 30 characters")
	}

	if !usernameRegex.MatchString(username) {
		errors = append(errors, "username must start with a letter and contain only letters, numbers, underscores, and hyphens")
	}

	reservedUsernames := []string{"admin", "root", "system", "api", "www", "mail", "support"}
	lowercaseUsername := strings.ToLower(username)
	for _, reserved := range reservedUsernames {
		if lowercaseUsername == reserved {
			errors = append(errors, "this username is reserved")
			break
		}
	}

	return errors
}

// Q: Should we implement password strength scoring for better user feedback?
func ValidatePassword(password string) []string {
	var errors []string

	if password == "" {
		errors = append(errors, "password is required")
		return errors
	}

	if len(password) < 8 {
		errors = append(errors, "password must be at least 8 characters")
	}

	hasUpper := false
	hasLower := false
	hasDigit := false
	hasSpecial := false

	for _, char := range password {
		switch {
		case unicode.IsUpper(char):
			hasUpper = true
		case unicode.IsLower(char):
			hasLower = true
		case unicode.IsDigit(char):
			hasDigit = true
		case unicode.IsPunct(char) || unicode.IsSymbol(char):
			hasSpecial = true
		}
	}

	if !hasUpper {
		errors = append(errors, "password must contain at least one uppercase letter")
	}

	if !hasLower {
		errors = append(errors, "password must contain at least one lowercase letter")
	}

	if !hasDigit {
		errors = append(errors, "password must contain at least one digit")
	}

	if !hasSpecial {
		errors = append(errors, "password must contain at least one special character")
	}

	// Check common passwords
	commonPasswords := []string{"password", "123456", "qwerty", "letmein", "welcome"}
	lowercasePassword := strings.ToLower(password)
	for _, common := range commonPasswords {
		if lowercasePassword == common {
			errors = append(errors, "this password is too common")
			break
		}
	}

	return errors
}

func ValidatePhone(phone string) []string {
	var errors []string

	if phone == "" {
		return errors // Phone is optional
	}

	// Remove common formatting characters
	cleaned := regexp.MustCompile(`[\s\-\(\)\.]`).ReplaceAllString(phone, "")

	if !phoneRegex.MatchString(cleaned) {
		errors = append(errors, "invalid phone number format")
	}

	return errors
}

func ValidateSlug(slug string) []string {
	var errors []string

	if slug == "" {
		errors = append(errors, "slug is required")
		return errors
	}

	if !slugRegex.MatchString(slug) {
		errors = append(errors, "slug must contain only lowercase letters, numbers, and hyphens")
	}

	return errors
}

type PasswordStrength int

const (
	PasswordWeak PasswordStrength = iota
	PasswordFair
	PasswordStrong
	PasswordVeryStrong
)

func GetPasswordStrength(password string) PasswordStrength {
	score := 0

	if len(password) >= 8 {
		score++
	}
	if len(password) >= 12 {
		score++
	}
	if len(password) >= 16 {
		score++
	}

	hasUpper := false
	hasLower := false
	hasDigit := false
	hasSpecial := false

	for _, char := range password {
		switch {
		case unicode.IsUpper(char):
			hasUpper = true
		case unicode.IsLower(char):
			hasLower = true
		case unicode.IsDigit(char):
			hasDigit = true
		case unicode.IsPunct(char) || unicode.IsSymbol(char):
			hasSpecial = true
		}
	}

	if hasUpper {
		score++
	}
	if hasLower {
		score++
	}
	if hasDigit {
		score++
	}
	if hasSpecial {
		score += 2
	}

	switch {
	case score >= 8:
		return PasswordVeryStrong
	case score >= 6:
		return PasswordStrong
	case score >= 4:
		return PasswordFair
	default:
		return PasswordWeak
	}
}

func (s PasswordStrength) String() string {
	switch s {
	case PasswordVeryStrong:
		return "very strong"
	case PasswordStrong:
		return "strong"
	case PasswordFair:
		return "fair"
	default:
		return "weak"
	}
}

File: tests/fixtures/vanilla-frontend/src/js/utils/validation.js
// src/js/utils/validation.js - Form validation utilities

export const validators = {
  required(value, message = 'This field is required') {
    if (value === null || value === undefined || value === '') {
      return message;
    }
    if (Array.isArray(value) && value.length === 0) {
      return message;
    }
    return null;
  },

  email(value, message = 'Invalid email address') {
    if (!value) return null;
    const pattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return pattern.test(value) ? null : message;
  },

  minLength(min, message) {
    return (value) => {
      if (!value) return null;
      if (value.length < min) {
        return message || `Must be at least ${min} characters`;
      }
      return null;
    };
  },

  maxLength(max, message) {
    return (value) => {
      if (!value) return null;
      if (value.length > max) {
        return message || `Must be at most ${max} characters`;
      }
      return null;
    };
  },

  pattern(regex, message = 'Invalid format') {
    return (value) => {
      if (!value) return null;
      return regex.test(value) ? null : message;
    };
  },

  // Q: Should we implement real-time password strength feedback?
  password(value, message = 'Password does not meet requirements') {
    if (!value) return null;

    const errors = [];

    if (value.length < 8) {
      errors.push('at least 8 characters');
    }
    if (!/[A-Z]/.test(value)) {
      errors.push('one uppercase letter');
    }
    if (!/[a-z]/.test(value)) {
      errors.push('one lowercase letter');
    }
    if (!/\d/.test(value)) {
      errors.push('one digit');
    }
    if (!/[!@#$%^&*(),.?":{}|<>]/.test(value)) {
      errors.push('one special character');
    }

    if (errors.length > 0) {
      return `Password must contain: ${errors.join(', ')}`;
    }
    return null;
  },

  match(fieldName, message) {
    return (value, formData) => {
      if (!value) return null;
      if (value !== formData[fieldName]) {
        return message || `Must match ${fieldName}`;
      }
      return null;
    };
  },

  phone(value, message = 'Invalid phone number') {
    if (!value) return null;
    const cleaned = value.replace(/[\s\-\(\)\.]/g, '');
    const pattern = /^\+?[0-9]{10,15}$/;
    return pattern.test(cleaned) ? null : message;
  },

  username(value, message = 'Invalid username') {
    if (!value) return null;

    if (value.length < 3) {
      return 'Username must be at least 3 characters';
    }
    if (value.length > 30) {
      return 'Username must be at most 30 characters';
    }
    if (!/^[a-zA-Z]/.test(value)) {
      return 'Username must start with a letter';
    }
    if (!/^[\w-]+$/.test(value)) {
      return 'Username can only contain letters, numbers, underscores, and hyphens';
    }

    const reserved = ['admin', 'root', 'system', 'api', 'www', 'mail', 'support'];
    if (reserved.includes(value.toLowerCase())) {
      return 'This username is reserved';
    }

    return null;
  },
};

export class FormValidator {
  constructor(rules) {
    this.rules = rules;
    this.errors = {};
  }

  validate(formData) {
    this.errors = {};
    let isValid = true;

    for (const [field, fieldRules] of Object.entries(this.rules)) {
      const value = formData[field];
      const fieldErrors = [];

      for (const rule of fieldRules) {
        const error = typeof rule === 'function' ? rule(value, formData) : null;
        if (error) {
          fieldErrors.push(error);
        }
      }

      if (fieldErrors.length > 0) {
        this.errors[field] = fieldErrors;
        isValid = false;
      }
    }

    return isValid;
  }

  getErrors() {
    return this.errors;
  }

  getFieldErrors(field) {
    return this.errors[field] || [];
  }

  hasErrors() {
    return Object.keys(this.errors).length > 0;
  }
}

export function getPasswordStrength(password) {
  if (!password) return { score: 0, label: 'None', color: 'gray' };

  let score = 0;

  if (password.length >= 8) score++;
  if (password.length >= 12) score++;
  if (password.length >= 16) score++;
  if (/[A-Z]/.test(password)) score++;
  if (/[a-z]/.test(password)) score++;
  if (/\d/.test(password)) score++;
  if (/[!@#$%^&*(),.?":{}|<>]/.test(password)) score += 2;

  const levels = [
    { min: 0, label: 'Weak', color: 'red' },
    { min: 4, label: 'Fair', color: 'orange' },
    { min: 6, label: 'Strong', color: 'blue' },
    { min: 8, label: 'Very Strong', color: 'green' },
  ];

  const level = levels.reverse().find((l) => score >= l.min) || levels[0];

  return { score, ...level };
}
